src/context.c

```diff
@@ -290,6 +291,189 @@ ly_ctx_new(const char *search_dir, uint16_t options, struct ly_ctx **new_ctx)
+    struct ly_set features = {0};
+    LY_ERR ret = LY_SUCCESS;
+    const struct lys_module *mod;
+    struct ly_ctx *ctx_yl, *ctx_new;
+
+    /* create a seperate context in case it is LY_CTX_NO_YANGLIBRARY since it needs it for parsing */
+    if (options & LY_CTX_NO_YANGLIBRARY) {
+        LY_CHECK_GOTO(ret = ly_ctx_new(search_dir, 0, &ctx_yl), error);
+        LY_CHECK_GOTO(ret = ly_ctx_new(search_dir, options, &ctx_new), error);
+    } else {
+        LY_CHECK_GOTO(ret = ly_ctx_new(search_dir, options, &ctx_yl), error);
+        ctx_new = ctx_yl;
+    }
+
+    /* parse yang library data tree */
+    LY_CHECK_GOTO(ret = parser_func(ctx_yl, input, format, 0, LYD_VALIDATE_PRESENT, &yltree), error);
+
+    LY_CHECK_GOTO(ret = lyd_find_xpath(yltree, "/ietf-yang-library:yang-library/module-set[1]/module", &set), error);
+
+    if (set->count == 0) {
+        /* perhaps a legacy data tree? */
+        LY_CHECK_GOTO(ret = ly_ctx_new_yl_legacy(ctx_new, yltree), error);
+    } else {
+        /* process the data tree */
+        for (uint32_t i = 0; i < set->count; ++i) {
+            module = set->dnodes[i];
+
+            /* initiate */
+            name = NULL;
+            revision = NULL;
+
+            /* Iterate over data */
+            LY_LIST_FOR(lyd_child(module), node) {
+                if (!strcmp(node->schema->name, "name")) {
+                    name = LYD_CANON_VALUE(node);
+                } else if (!strcmp(node->schema->name, "revision")) {
+                    revision = LYD_CANON_VALUE(node);
+                } else if (!strcmp(node->schema->name, "feature")) {
+                    LY_CHECK_GOTO(ret = ly_set_add(&features, node, 0, NULL), error);
+                }
+            }
+
+            feature_arr = malloc(sizeof *feature_arr * (features.count + 1));
+            LY_CHECK_ERR_GOTO(!feature_arr, ret = LY_EMEM, error);
+
+            /* Parse features into an array of strings */
+            for (uint32_t u = 0; u < features.count; u++) {
+                feature_arr[u] = LYD_CANON_VALUE(features.dnodes[u]);
+            }
+            feature_arr[features.count] = NULL;
+            ly_set_clean(&features, NULL);
+
+            /* use the gathered data to load the module */
+            mod = ly_ctx_load_module(ctx_new, name, revision, feature_arr);
+            free(feature_arr);
+            feature_arr = NULL;
+            if (!mod) {
+                LOGERR(NULL, LY_EINVAL, "Unable to load module specified by yang library data.");
+                ret = LY_EINVAL;
+                goto error;
+            }
+        }
+    }
+
+error:
+
+    /* cleanup */
+    /* yang library data tree */
+    lyd_free_all(yltree);
+    ly_set_free(set, NULL);
+    ly_set_erase(&features, NULL);
+    if (feature_arr) {
+        free(feature_arr);
+    }
+    if (ctx_yl != ctx_new) {
+        ly_ctx_destroy(ctx_yl, NULL);
+    }
+    *ctx = ctx_new;
+    if (ret) {
+        ly_ctx_destroy(*ctx, NULL);
+        *ctx = NULL;
+    }
```

Comment:
Just a minor improvement, I would write it like
```C
    if (ret) {
        ly_ctx_destroy(ctx_new, NULL);
        *ctx = NULL;
    } else {
        *ctx = ctx_new;
    }
```
which clearly shows the difference between success and error. But you can keep it this way.